Feat/issue 5784 custom max tokens

MuriloFP opened the pull request:
PR Title: feat: Add Advanced Setting for Custom Max Tokens per Provider Profile (#5784)

---

<!--
Thank you for contributing to Roo Code!

Before submitting your PR, please ensure:
- It's linked to an approved GitHub Issue.
- You've reviewed our [Contributing Guidelines](../CONTRIBUTING.md).
-->

### Related GitHub Issue

<!-- Every PR MUST be linked to an approved issue. -->

Closes: #5784 <!-- Replace with the issue number, e.g., Closes: #123 -->

### Roo Code Task Context (Optional)

<!--
If you used Roo Code to help create this PR, you can share public task links here.
This helps reviewers understand your development process and provides additional context.
Example: https://app.roocode.com/share/task-id
-->

### Description

<!--
Briefly summarize the changes in this PR and how they address the linked issue.
The issue should cover the "what" and "why"; this section should focus on:
- The "how": key implementation details, design choices, or trade-offs made.
- Anything specific reviewers should pay attention to in this PR.
-->

This PR adds a new "Max Output Tokens" field in the Advanced Settings section of the provider configuration UI, allowing users to customize the max tokens per provider profile. Previously, Roo Code had a hard-coded limit of 8192 tokens for all API providers, which prevented users from fully utilizing models that support higher token limits.

**Key implementation details:**
- Created a new `MaxTokensControl` component with numeric input validation
- Integrated the control into the Advanced Settings section of `ApiOptions`
- Updated `getModelMaxOutputTokens()` to respect user-configured `modelMaxTokens` for all models
- Added comprehensive translations for all supported locales (23 languages)
- Maintained backward compatibility with a default of 8192 tokens

**Design choices:**
- Minimum value set to 1000 tokens (reasonable minimum for practical use)
- Maximum value capped at model's actual max tokens or 200,000
- Validation errors displayed inline with clear messages
- Model's supported max tokens shown as a hint to guide users

### Test Procedure

<!--
Detail the steps to test your changes. This helps reviewers verify your work.
- How did you test this specific implementation? (e.g., unit tests, manual testing steps)
- How can reviewers reproduce your tests or verify the fix/feature?
- Include relevant testing environment details if applicable.
-->

**Automated Tests:**
- ✅ Backend API tests: 22 tests passing in `src/shared/__tests__/api.spec.ts`
- ✅ UI Component tests: 10 tests passing in `webview-ui/src/components/settings/__tests__/MaxTokensControl.spec.tsx`
- ✅ All ESLint and TypeScript checks passing

**Manual Testing Steps:**
1. Navigate to Settings > API Provider configuration
2. Select any provider (e.g., Anthropic, OpenAI)
3. Expand the "Advanced Settings" section
4. Verify the "Max Output Tokens" field appears with default value of 8192
5. Enter a custom value (e.g., 16000) and save
6. Switch to a different provider profile and verify it maintains its own value
7. Try invalid values (< 1000 or > model max) and verify validation errors appear
8. Use the provider with custom max tokens and verify it's applied in API requests

### Pre-Submission Checklist

<!-- Go through this checklist before marking your PR as ready for review. -->

- [x] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [x] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [x] **Self-Review**: I have performed a thorough self-review of my code.
- [x] **Testing**: New and/or updated tests have been added to cover my changes (if applicable).
- [x] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [x] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### Screenshots / Videos

<!--
For UI changes, please provide before-and-after screenshots or a short video of the *actual results*.
This greatly helps in understanding the visual impact of your changes.
-->

[Screenshots should be added showing the new Max Output Tokens field in the Advanced Settings section]

### Documentation Updates

<!--
Does this PR necessitate updates to user-facing documentation?
- [ ] No documentation updates are required.
- [ ] Yes, documentation updates are required. (Please describe what needs to be updated or link to a PR in the docs repository).
-->

- [ ] Yes, documentation updates are required. The settings documentation should be updated to explain the new "Max Output Tokens" field and its impact on token distribution.

### Additional Notes

<!-- Add any other context, questions, or information for reviewers here. -->

This implementation leverages the existing `modelMaxTokens` field in `ProviderSettings` which was previously only used for reasoning models. The field is now extended to work with all models, providing a consistent experience across different provider types.

### Get in Touch

<!--
Please provide your Discord username for reviewers or maintainers to reach you if they have questions about your PR
-->

@MuriloFP 

<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Introduces customizable max output tokens per provider profile in the UI, with backend support and comprehensive testing.
> 
>   - **Behavior**:
>     - Adds "Max Output Tokens" field in provider configuration UI for custom token limits.
>     - Default limit remains 8192 tokens; users can set between 1000 and model's max or 200,000.
>     - Validation errors for invalid values are shown inline.
>   - **Components**:
>     - New `MaxTokensControl` component with numeric input validation.
>     - Integrated into `ApiOptions` in `ApiOptions.tsx`.
>   - **Functions**:
>     - Updates `getModelMaxOutputTokens()` to use user-configured `modelMaxTokens`.
>     - Applies changes in `chutes.ts`, `gemini.ts`, and `glama.ts`.
>   - **Testing**:
>     - 22 backend API tests in `api.spec.ts`.
>     - 10 UI tests in `MaxTokensControl.spec.tsx`.
>   - **Translations**:
>     - Adds translations for 23 languages.
>   - **Misc**:
>     - Maintains backward compatibility with default 8192 tokens.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for 8ade97f0f466148d5b473385c9bd8114d4d6aaf5. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

daniel-lxs reviewed (commented):
Hey @MuriloFP, I left a couple of minor suggestions, the implementation looks good overall

Let me know what you think!

daniel-lxs commented:
It seems like I can’t completely clear the field because as soon as I delete everything, the default value gets restored. This is likely due to the check happening in the `onChange` event. It would be better to move that logic to the `onBlur` event so users have a chance to enter a new value before validation kicks in.

MuriloFP commented:
Hi @daniel-lxs,

I've addressed your feedback about the input field clearing issue. The fix has been implemented and pushed in commit e6cc00bd.

## Changes Made

### Fixed Input Clearing Behavior ✅
- The Max Output Tokens field now allows users to clear it completely without immediately restoring the default value
- Default value is only applied when the user leaves the field empty (on blur event)
- Users can now clear the field and type a new value without interruption

### Implementation Details
- Added local state management to track the input value separately
- Moved default value application from `onChange` to `onBlur` handler
- Updated validation to not show "too low" error for empty fields during editing
- The field shows empty state while typing, making it clear that the user can enter any value

### Tests Updated
- All tests have been updated to reflect the new behavior
- Added specific tests for blur behavior and empty field handling
- All 14 MaxTokensControl tests passing ✅
- All 13 ApiOptions tests passing ✅

The user experience is now much smoother - users can clear the field, see it empty, type their desired value, and only see the default applied if they leave the field completely empty.

Let me know if this addresses your concern or if you'd like any adjustments!

daniel-lxs reviewed (commented):
I think we should discuss how to handle reasoning vs non-reasoning models since reasoning models already have a slider for Max output tokens

<img width="365" height="520" alt="Image" src="https://github.com/user-attachments/assets/85a22b05-37be-43cf-b533-0b78ef394459" />

Some allowed values seem to mess up the Max Thinking Tokens value as well:

<img width="365" height="520" alt="Image" src="https://github.com/user-attachments/assets/8b5d6406-0125-4ccd-939f-92880072d0bf" />

How do you think we should handle this? Should we remove the Max Tokens slider?

mrubens commented:
Hmm, not sure the setting is sticking correctly for me

<img width="463" height="1172" alt="Screenshot 2025-07-30 at 10 27 59 PM" src="https://github.com/user-attachments/assets/2da10789-50b7-4b58-aab1-ad545593b6da" />
<img width="461" height="332" alt="Screenshot 2025-07-30 at 10 28 15 PM" src="https://github.com/user-attachments/assets/462e059c-f565-462c-844c-1d4d4218abb0" />

MuriloFP commented:
> Hmm, not sure the setting is sticking correctly for me
> 

The providers I hadn't tested are getting the max output from the model directly, I need to change them all... The commit above fixed it for Chutes, but I still need to change the other occurrences for the other providers. Marked this as draft, will get it done ASAP, but I don't have access to all of the providers, will test the ones I can.

MuriloFP commented:
@mrubens 
I've made changes to the following providers:

Chutes
Glama
Mistral
xAI
Gemini
Unbound
LiteLLM

I've tested all of them except for Unbound and LiteLLM, they all seem to be working fine. Previously, they were getting the max tokens directly from the model, now it uses the custom setting.

mechanicmuthu commented:
What happens if user setting of model max is decreased in between a conversation and the current context is more than the user setting?
